v2/bundle/spiffebundle/bundle.go

```diff
@@ -3,175 +3,411 @@ package spiffebundle
 import (
 	"crypto"
 	"crypto/x509"
+	"encoding/json"
 	"io"
+	"io/ioutil"
+	"sync"
 	"time"
 
 	"github.com/spiffe/go-spiffe/v2/bundle/jwtbundle"
 	"github.com/spiffe/go-spiffe/v2/bundle/x509bundle"
+	"github.com/spiffe/go-spiffe/v2/internal/x509util"
 	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/zeebo/errs"
+	"gopkg.in/square/go-jose.v2"
 )
 
+const (
+	x509SVIDUse = "x509-svid"
+	jwtSVIDUse  = "jwt-svid"
+)
+
+var (
+	spiffebundleErr = errs.Class("spiffebundle")
+)
+
+type bundleDoc struct {
+	jose.JSONWebKeySet
+	SequenceNumber uint64 `json:"spiffe_sequence,omitempty"`
+	RefreshHint    int64  `json:"spiffe_refresh_hint,omitempty"`
+}
+
 // Bundle is a collection of trusted public key material for a trust domain,
-// confirming to the SPIFFE Bundle Format as part of the SPIFFE Trust Domain
+// conforming to the SPIFFE Bundle Format as part of the SPIFFE Trust Domain
 // and Bundle specification:
 // https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Trust_Domain_and_Bundle.md
 type Bundle struct {
+	trustDomain spiffeid.TrustDomain
+
+	mtx            sync.RWMutex
+	refreshHint    *int64
```

Comment:
I think things are a little nicer if this is a `time.Duration` and the conversion happens during marshalling/unmarshalling instead of inside SetRefreshHint/RefreshHint.